### PR TITLE
Fixes #16542 - More thorough pulp services check

### PR DESCRIPTION
**Pull request by johnpmitsch**
pulp should be checked to be fully functional by checking

```
verify messaging_connection is True
verify database_connection is True
verify at least one scheduler@ entry is present
verify at least one resource_manager@ entry is present
verify at least one reserved_resource_worker-*@ is present
```

### CODE REVIEW

File: app/models/katello/ping.rb [hunk 1]
```diff
@@ -129,8 +125,24 @@ def pulp_without_auth(url)
...
+        unless all_pulp_workers_present?(json)
+          fail _("Not all necessary pulp workers running.")
+        end
+
         json
       end
+
+      def all_pulp_workers_present?(json)
+        worker_ids = json["known_workers"].collect { |worker| worker["_id"] }
+        if worker_ids
```
Comment (daviddavis) [inline app/models/katello/ping.rb]: I would probably use a [guard clause](https://git.io/vi25e) here. I think it'll simplify the code.

**Comment by daviddavis** [comment]
Any way to add a test for this change?

File: app/models/katello/ping.rb (hunk 1 above)
Comment (daviddavis) [inline app/models/katello/ping.rb]: Actually, `collect` always returns a non-nil value. So this check will always evaluate to true.

**Comment by ehelms** [comment]
Has a Pulp dev looked at this? I seem to recall a conversation once where someone indicated we werent checking or interpreting their status properly.

**Comment by johnpmitsch** [comment]
@daviddavis good call, we should be testing this
@ehelms I talked with @bmbouter about it on #pulp, he may want to take a look at this to ensure I am understanding him correctly

File: app/models/katello/ping.rb [hunk 2]
```diff
@@ -129,8 +125,21 @@ def pulp_without_auth(url)
...
+        unless all_pulp_workers_present?(json)
+          fail _("Not all necessary pulp workers running.")
+        end
+
         json
       end
+
+      def all_pulp_workers_present?(json)
+        worker_ids = json["known_workers"].collect { |worker| worker["_id"] }
+        return false unless worker_ids.any?
```
Comment (johnpmitsch) [inline app/models/katello/ping.rb]: @daviddavis updated to check if anything is present in the array returned

**Comment by johnpmitsch** [comment]
@daviddavis added tests :100:

**Comment by johnpmitsch** [comment]
To test:

check each pulp service being shut off individually causes failing ping
SERVICES = [pulp_resource_manager.service, pulp_workers.service, pulp_celerybeat.service]
For service in SERVICES:
1) ```systemctl stop service```
2) check Katello::Ping.ping fails from the rails console
3) ```systemctl start service```

With all services turned on, check Katello::Ping.ping works from the console
With all services turned off, check Katello::Ping.ping works from the console

Then its worth testing basic pulp functionality like repo syncing and CV publishes

**Comment by daviddavis** [comment]
Do we want to have @bmbouter sign off on this before we merge?

**Comment by bmbouter** [comment]
I can't speak to if the ruby is correct or not, but I can see each required check has a line which reads like it is doing the right thing. This looks good to merge IMO.

Thank you all so much for doing this!

**Pull request merged**